sqlgen/integration_test.go

```diff
@@ -174,6 +196,178 @@ func TestBatchFilter(t *testing.T) {
+					return err
+				}
+				return nil
+			},
+			goroutineExpectedResult: []*Owl{
+				{Species: "Bubo bubo", CommonName: "Eurasian Eagle-Owl", Genus: "Bubo", Family: "Strigidae"},
+				{Species: "Bubo virginianus", CommonName: "Great Horned Owl", Genus: "Bubo", Family: "Strigidae"},
+			},
+		},
+		{
+			name:              "main thread blocks with broad query",
+			description:       "Main thread performs a broad SELECT ... FOR UPDATE. The goroutine does a narrow SELECT ... FOR UPDATE and should be blocked by the main thread's query. The main thread INSERTs a new row, and the goroutine should pick it up.",
+			mainThreadFilter:  Filter{"family": "Strigidae"},
+			mainThreadOptions: &SelectOptions{ForUpdate: true},
+			goroutineFilter:   Filter{"family": "Strigidae", "genus": "Megascops"},
+			goroutineOptions:  &SelectOptions{ForUpdate: true, OrderBy: "species"},
+			mainThreadUpdateFunc: func(ctx context.Context, db *DB, waiter *sync.WaitGroup) error {
+				// INSERT A new row that the goroutine should get in its SELECT query
+				if _, err := db.InsertRow(ctx, &Owl{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"}); err != nil {
+					return err
+				}
+				return nil
+			},
+			goroutineExpectedResult: []*Owl{
+				{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"},
+				{Species: "Megascops kennicottii", CommonName: "Western Screech-Owl", Genus: "Megascops", Family: "Strigidae"},
+			},
+		},
+		{
+			name:              "non-blocking queries on different parts of the index",
+			description:       "The main thread and the goroutine both perform a SELECT ... FOR UPDATE, but their queries shouldn't block each other because they are selecting different parts of the index",
+			mainThreadFilter:  Filter{"family": "Strigidae", "genus": "Megascops"},
+			mainThreadOptions: &SelectOptions{ForUpdate: true},
+			goroutineFilter:   Filter{"family": "Strigidae", "genus": "Psiloscops"},
+			goroutineOptions:  &SelectOptions{ForUpdate: true},
+			mainThreadUpdateFunc: func(ctx context.Context, db *DB, waiter *sync.WaitGroup) error {
+				// Wait until the goroutine has performed its SELECT
+				waiter.Wait()
+
+				// INSERT a new row
+				if _, err := db.InsertRow(ctx, &Owl{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"}); err != nil {
+					return err
+				}
+				return nil
+			},
+			goroutineExpectedResult: []*Owl{
+				{Species: "Psiloscops flammeolus", CommonName: "Flammulated Owl", Genus: "Psiloscops", Family: "Strigidae"},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create the test database and insert some initial rows into the owls table
+			tdb, db, err := setup()
+			assert.NoError(t, err)
```

Comment:
`setup()` failing is a fatal error. Would you use `require` instead?

Reply:
Ah yeah, a lot of these `assert`s should actually stop the test.  I'll update many of them to be `require` instead of `assert`.